happi/containers.py

```diff
@@ -30,25 +86,30 @@ def __new__(cls, *args, **kwargs):
             cls.__instance.__initialized = False
         return cls.__instance
 
-    def __getitem__(self, item):
-        if item not in self._registry:
+    def __getitem__(self, item: str) -> Optional[Type[HappiItem]]:
+        if not self._loaded:
             self.load()
         return self._registry.get(item)
 
-    def __contains__(self, item):
-        if item not in self._registry:
+    def __setitem__(self, item: str, klass: Type[HappiItem]) -> None:
+        self._safe_add(item, klass)
+
+    def __contains__(self, item: str):
+        if not self._loaded:
             self.load()
         return item in self._registry
 
-    def items(self):
+    def items(self) -> Generator[Tuple[str, Type[HappiItem]], None, None]:
+        if not self._loaded:
+            self.load()
         yield from self._registry.items()
 
     def entry_for_class(self, klass):
-        if klass not in self._reverse_registry:
+        if not self._loaded:
```

Comment:
Switching to this conditional means that we don't try to reload the registry if we can't find a class, right?  That seems like nice functionality to have, though it seems rare to get a new container at runtime.

Reply:
That's right. I think the original implementation isn't quite right, really. The edge cases I can imagine are either not really possible to hit or ones we might not want to support:

1. Someone mutates the `DEFAULT_REGISTRY` in `happi.containers` and hopes it'll be picked up after `__getitem__` falls through
2. Someone monkeypatches entrypoints to pick up a new set of containers (this could be fine in a test suite, but just call `.load()` if you need to do this)
3. One I'm unsure of: can you launch `ipython`, import happi, and in a separate terminal do a `pip install` and see updated entrypoints in the original Python instance? I think the answer is "probably not". And even if so, do we care to support this?

Reply:
In the interest of keeping things more or less the same as the master branch - whether or not the external `pip install` stuff makes sense to support - I've added that in and will mark this resolved